pkg/dataloader/jiraloader/jiraloader.go

```diff
@@ -0,0 +1,354 @@
+					if err != nil {
+						log.WithError(err).Warningf("parsing error: %s", history.Created)
+						continue
+					}
+					// We pick the oldest state change
+					if resolutionTime == nil || resolutionTime.After(createdTime) {
+						log.Debugf("%s to %s at %+v", issue.Key, status, createdTime)
+						resolutionTime = &createdTime
+					}
+				}
+			}
+		}
+	}
+
+	// Fallback to the jira resolution time
+	if issue.Fields.ResolutionDate != "" && resolutionTime == nil {
+		jiraResolutionTime, err := time.Parse(jiraTimeLayout, issue.Fields.ResolutionDate)
+		if err != nil {
+			fmt.Printf("parsing error: %+v", err)
+		}
+		log.Debugf("resolution time for %s is %+v", issue.Key, jiraResolutionTime)
+		resolutionTime = &jiraResolutionTime
+	}
+
+	return resolutionTime
+}
+
+func issueContainsLabel(issue *v1jira.Issue, label string) bool {
+	for _, issueLabel := range issue.Fields.Labels {
+		if issueLabel == label {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (jl *JiraLoader) componentLoader() {
+	start := time.Now()
+	log.Infof("loading jira ocpbugs component information...")
+	body, err := jiraRequest("https://issues.redhat.com/rest/api/2/project/12332330/components")
+	if err != nil {
+		jl.errors = append(jl.errors, err)
+		return
+	}
+
+	var components []v1jira.Component
+	err = json.Unmarshal(body, &components)
+	if err != nil {
+		jl.errors = append(jl.errors, err)
+		return
+	}
+
+	tblUpdater := stagingtablemanager.New(jl.dbc, "jira_components")
+	tmpTable, err := tblUpdater.CreateStagingTable()
+	if err != nil {
+		jl.errors = append(jl.errors, errors.Wrap(err, "couldn't create temporary table"))
+		return
+	}
+	defer func() {
+		if err := tblUpdater.Cleanup(); err != nil {
+			jl.errors = append(jl.errors, errors.Wrap(err, "couldn't cleanup table updater"))
+			return
+		}
+	}()
+
+	for _, c := range components {
+		jiraID, err := strconv.ParseUint(c.ID, 10, 64)
+		if err != nil {
+			msg := "error parsing jira ID"
+			log.WithError(err).Warn(msg)
+			jl.errors = append(jl.errors, errors.WithMessage(err, msg))
+			continue
+		}
+
+		mc := models.JiraComponent{
+			Model: models.Model{
+				ID: uint(jiraID),
+			},
+			Name:        c.Name,
+			Description: c.Description,
+			LeadName:    c.Lead.DisplayName,
+			LeadEmail:   c.Lead.Name,
+		}
+
+		// Save to main table
+		if err := jl.dbc.DB.Clauses(clause.OnConflict{UpdateAll: true}).Save(&mc).Error; err != nil {
+			jl.errors = append(jl.errors, err)
+			log.WithError(err).Warningf("failed to save component %q", c.Name)
+			continue
+		}
+
+		// Save to staging table
+		res := jl.dbc.DB.Table(tmpTable).Save(&mc)
+		if res.Error != nil {
+			jl.errors = append(jl.errors, res.Error)
+			log.WithError(err).Warningf("failed to save component %q", c.Name)
+			return
+		}
+
+		fmt.Println(mc.ID)
+	}
+
+	log.Infof("deleting old records...")
+	oldRecords := jl.dbc.DB.Exec(fmt.Sprintf(`DELETE FROM jira_components WHERE id NOT IN (SELECT id FROM %s)`, tmpTable))
```

Comment:
How many records in tmpTable today? This will break at a certain point with postgres but hopefully we're far away. 64k items IIRC?

Reply:
What are the advantages to the staging table vs maintaining a list of what you expect in memory and using it in the query?

Reply:
> How many records in tmpTable today? This will break at a certain point with postgres but hopefully we're far away.

What would break with this particular query? This is designed to not give a list of thousands of ID's directly.

> What are the advantages to the staging table vs maintaining a list of what you expect in memory and using it in the query?

Yea it's around 64K items.  That itself isn't a big deal, I couldn't find good information about what the maximum query size was since it's dependent on a bunch of factors.  I was just thinking in general about our future needs to sync data from bigquery.  Most of the docs online recommend using temporary tables like this and doing upserts for large syncs.

I've gone back and forth on this a bunch.  Maybe it's worth worrying about when we have more data to sync, I pushed up the id-based version if you prefer that for now